mut/index/SnootyManifest.py

```diff
@@ -0,0 +1,228 @@
+from bson import decode_all
+from jsonpath_ng.ext import parse
+from os import walk
+from os.path import join
+from pathlib import Path
+from json import dumps
+from concurrent.futures import ProcessPoolExecutor
+import logging
+
+from typing import Optional, List, Tuple, TypedDict
+
+logger = logging.getLogger(__name__)
+
+
+class ManifestEntry(TypedDict):
+    slug: str
+    title: Optional[str]
+    headings: Optional[List[str]]
+    paragraphs: str
+    code: dict
+    preview: Optional[str]
+    tags: list[str]
+
+
+class Document:
+    """Return indexing data from a page's AST for search purposes."""
+
+    def __init__(self, data) -> None:
+
+        self.tree = data[0]
+
+        self.robots, self.keywords, self.description = self.find_metadata()
+
+        self.paragraphs = self.find_paragraphs()
+        self.code = self.find_code()
+        self.title, self.headings = self.find_headings()
+        self.slug = self.derive_slug()
+        self.preview = self.derive_preview()
+
+        self.noindex, self.reasons = self.get_noindex()
+
+    def find_paragraphs(self) -> str:
+        logger.debug("Finding paragraphs")
+        # NB: paragraphs include "paragraph" nodes within tables
+        jsonpath_expr = parse("$..children[?(@.type=='paragraph')]..value")
+        results = jsonpath_expr.find(self.tree)
+
+        # Appending to then joining an array is faster than repeatedly concatenating strings
+        str_list = []
```

Comment:
you could idiomatically do this as a list comprehension:
```
str_list = [r.value for r in results]
```

Reply:
that is slick, but I am also ready to close this out :D thanks for the learn <3